core/shared/platform/windows/win_file.c

```diff
@@ -1392,15 +1708,38 @@ os_readdir(os_dir_stream dir_stream, __wasi_dirent_t *entry,
 {
     CHECK_VALID_WIN_DIR_STREAM(dir_stream);
 
-    return __WASI_ENOSYS;
+    FILE_ID_BOTH_DIR_INFO *file_id_both_dir_info = NULL;
+
+    __wasi_errno_t error =
+        read_next_dir_entry(dir_stream, &file_id_both_dir_info);
+
+    if (error != __WASI_ESUCCESS || file_id_both_dir_info == NULL)
+        return error;
+
+    entry->d_ino = (__wasi_inode_t)file_id_both_dir_info->FileId.QuadPart;
+    entry->d_namlen = (__wasi_dirnamlen_t)(file_id_both_dir_info->FileNameLength
+                                           / (sizeof(wchar_t) / sizeof(char)));
+    entry->d_next = (__wasi_dircookie_t)dir_stream->cookie;
+    entry->d_type = get_disk_filetype(file_id_both_dir_info->FileAttributes);
+
+    *d_name = dir_stream->current_entry_name;
+
+    return __WASI_ESUCCESS;
 }
 
 __wasi_errno_t
 os_closedir(os_dir_stream dir_stream)
 {
     CHECK_VALID_WIN_DIR_STREAM(dir_stream);
 
-    return __WASI_ENOSYS;
+    bool success = CloseHandle(dir_stream->handle->raw.handle);
+
+    if (!success)
+        return convert_windows_error_code(GetLastError());
```

Comment:
Should we `BH_FREE(dir_stream)` before return error?

Reply:
I don't think so - it's hard to know from the [CloseHandle documentation](https://learn.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle) but if the error returned is recoverable, we would leak the handle if it isn't closed here and we free the owning `dir_stream` struct. In practice, I think the only reason `CloseHandle` would fail is if the provided handle is invalid - so we should probably call `BH_FREE(dir_stream)` if `GetLastError` returns `ERROR_INVALID_HANDLE`.